Add daemonsets

ekimekim opened the pull request:
Fixes https://github.com/weaveworks/scope/issues/1444

This adds collection of daemonset info to the probe,
and two new views to the app. One showing only daemonsets,
one showing deployments + daemonsets.

This is too many views (see discussion in #1444), but was useful for initial debugging - we can remove the daemonsets view before merging.

We need to decide how we are going to present this new 'combined view' to the user in a way that makes it obvious what's going on. For now I've labelled it 'Replica groups' but that's not good enough.

2opremio commented on probe/kubernetes/daemonset.go:
```diff
@@ -0,0 +1,53 @@
+	selector, err := unversioned.LabelSelectorAsSelector(d.Spec.Selector)
+	if err != nil {
+		// TODO(paulbellamy): Remove the panic!
+		panic(err)
```
Can you propagate the error instead of the TODO? :)

2opremio commented on render/pod.go:
```diff
@@ -96,6 +96,28 @@ var ReplicaSetRenderer = ConditionalRenderer(renderKubernetesTopologies,
+// ReplicaGroupRenderer is a Renderer which combines the daemonset and deployment views
+// in (for now) a very naive way.
+var ReplicaGroupRenderer = MakeReduce(
+	DeploymentRenderer,
```
I am not sure I get this. The ReplicaGroup view will also include DaemonSets? Don't we have separate view for that?

That's confusing ...

2opremio commented on render/pod.go:
```diff
@@ -96,6 +96,28 @@ var ReplicaSetRenderer = ConditionalRenderer(renderKubernetesTopologies,
+// in (for now) a very naive way.
+var ReplicaGroupRenderer = MakeReduce(
+	DeploymentRenderer,
+	DaemonSetRenderer,
```
So, the ReplicaGroup view will also include DaemonSets? Why have a separate view for DaemonSets alone then?

Also, I am not sure I like the name (ReplicaGroup) or the entities it groups. 

I think it would make sense to group controllers together (that is, ReplicaSets, ReplicationControllers, DaemonSets, PetSets) in a view and name it  "Controllers". However, Deployments live one level of abstraction higher from controllers (they are on top of ReplicaSets) which also makes the name confusing.

2opremio commented:
Done with the review: Minor code comment and a fundamental design question.

I don't mind going ahead with this a refining it later, but as a kubernetes user I find it a bit odd.

ekimekim commented:
(addressing design question at the top level instead of at the specific comment)

Yeah, I agree. The view is confusing and right now badly named. `Controllers` is a better name (though I'm unsure if it's a good name either). I added the daemonset view itself because it was trivial to do and let me ensure it was working correctly.

The `ReplicaGroup` view was an attempt at addressing concerns in #1444, and yeah I'm looking for ideas on how to make this less confusing.

As I see it, we have three options. In my order of preference:
1. Drop the combined view from this PR and merge it with just the seperate views, as was the original intent of #1444. We can then work on a better, non-rushed version of the combined view.
2. Merge this PR as-is, with both daemonset view and combined view, and iterate on these issues in a followup.
3. Delay merging this PR (which adds any support for daemonsets at all) until the combined issue is resolved.

2opremio commented:
I would go for (1) as well. @rade ?

rade commented:
go for (1), but make sure there is a follow up, in particular peeling off bits from #552, specifically replacing (nearly) all existing views with a view that shows you everything in existence, at the highest level of abstraction in which it has a representation.

ekimekim commented:
I've removed the combined view parts and otherwise rebased fixups. PTAL.

2opremio commented:
Looks good. There is still a comment on a TODO

ekimekim commented:
Rebased and resolved conflicts with https://github.com/weaveworks/scope/pull/2528.
Now propagates errors from getting label selectors for all types, not just daemonsets.
PTAL.

2opremio commented:
I am exceptionally merging this myself (and not the PR creator) to move forward with the release
